### PR TITLE
:bug: Populate GVK for listed objects

### LINKED ISSUE #284: MapObject.Object contains empty GroupVersionKind
I noticed that `MapObject.Object` instances passed to a `handler.ToRequestsFunc` function contains empty `GroupVersionKind`. The code I have looks like the following:

```go
mapFn := handler.ToRequestsFunc(
	func(a handler.MapObject) []reconcile.Request {
		var requests []reconcile.Request
		apps := listApplications(mgr.GetClient(), a.Meta.GetNamespace())
		for _, app := range apps {
			ref := app.Spec.Selector.ObjectRef
			gvk := a.Object.GetObjectKind().GroupVersionKind()
			if gvk.Empty() {
				log.Printf("Empty GVK for object %s/%s", a.Meta.GetNamespace(), a.Meta.GetName())
				return requests
			}
			if ref != nil && ref.Name == a.Meta.GetName() && ref.Kind == gvk.Kind {
				requests = append(requests, reconcile.Request{
					NamespacedName: types.NamespacedName{
						Namespace: app.Namespace,
						Name:      app.Name,
					},
				})
			}
		}
		return requests
	})
``` 

The logs show multiple occurrences of `Empty GVK for object ...`.

### PR DESCRIPTION
**Pull request by liyinan926**
Fixes #284. Populates `GroupVersionKind` for objects returned by the `List` method.

### CODE REVIEW

File: pkg/cache/cache_test.go [hunk 1]
```diff
@@ -198,6 +197,19 @@ var _ = Describe("Informer Cache", func() {
...
+			It("should be able to list objects with GVK populated", func() {
+				By("listing pods")
+				listObj := &kcorev1.PodList{}
+				Expect(informerCache.List(context.Background(), listObj)).To(Succeed())
+
+				By("verifying that the returned pods have GVK populated")
+				Expect(listObj.Items).NotTo(BeEmpty())
+				Expect(listObj.Items).Should(HaveLen(3))
+				for _, p := range listObj.Items {
+					Expect(p.GroupVersionKind().Empty()).To(BeFalse())
```
Comment (DirectXMan12) [inline pkg/cache/cache_test.go]: this should test that we list the right gvk (the `corev1` package should have the package group-version as a package-level variable)

Comment (liyinan926) [reply pkg/cache/cache_test.go]: Done.

**Comment by liyinan926** [comment]
Is this good to merge now?

File: pkg/cache/cache_test.go [hunk 2]
```diff
@@ -198,6 +197,21 @@ var _ = Describe("Informer Cache", func() {
...
+				listObj := &kcorev1.PodList{}
+				Expect(informerCache.List(context.Background(), listObj)).To(Succeed())
+
+				By("verifying that the returned pods have GVK populated")
+				Expect(listObj.Items).NotTo(BeEmpty())
+				Expect(listObj.Items).Should(HaveLen(3))
+				for _, p := range listObj.Items {
+					Expect(p.GroupVersionKind().Group).To(Equal(kcorev1.SchemeGroupVersion.Group))
+					Expect(p.GroupVersionKind().Version).To(Equal(kcorev1.SchemeGroupVersion.Version))
+					Expect(p.GroupVersionKind().Kind).To(Equal("Pod"))
```
Comment (DirectXMan12) [inline pkg/cache/cache_test.go]: `Expect(p.GroupVersionKind()).To(Equal(kcorev1.SchemeGroupVersion.WithKind("Pod")))`

Comment (liyinan926) [reply pkg/cache/cache_test.go]: Done.

**Comment by DirectXMan12** [comment]
Looks good.  Please squash the commits into logical commits (i.e. no "address comments" commits), and I'll merge.

**Comment by shawn-hurley** [comment]
@liyinan926 think you could rebase this pr?

**Comment by shawn-hurley** [comment]
@liyinan926 do you think that you can rebase this PR?

**Comment by JoelSpeed** [comment]
Just run into this issue, @liyinan926 have you time to fix the conflicts and rebase the PR? Happy to pick this up if you don't

**Comment by JoelSpeed** [comment]
This is blocking me on a project so I've rebased the branch so that I can continue, let me know if you'd like me to open a PR to superseed this one https://github.com/pusher/controller-runtime/tree/list-gvk

I found an issue with the tests in which the length of the list can be three or four depending on which cache type we are using. For now I've added a `SatisfyAny(HaveLen(3), HaveLen(4))` to the matcher, any ideas for a better way to fix this?

**Comment by JoelSpeed** [comment]
I've also just had to push a change to `DeepCopyObject()` the objects before fixing their GVKs as it was setting off the race detector in my tests, looks like listing stopped deep copying at some point recently?

**Comment by DirectXMan12** [comment]
@JoelSpeed go ahead and pick this up, and just credit @liyinan926 / keep them as an author on the commit.

**Comment by JoelSpeed** [comment]
@DirectXMan12 / @liyinan926 I've opened #389 from my branch keeping original commit authorship to credit @liyinan926

**Comment by droot** [comment]
closing this since these changes by pull in a separate PR https://github.com/kubernetes-sigs/controller-runtime/pull/389

**Pull request closed without merging**